gensim/models/fasttext.py

```diff
@@ -0,0 +1,374 @@
+    """
+    Update skip-gram model by training on a sequence of sentences.
+
+    Each sentence is a list of string tokens, which are looked up in the model's
+    vocab dictionary. Called internally from `Word2Vec.train()`.
+
+    This is the non-optimized, Python version. If you have cython installed, gensim
+    will use the optimized version from word2vec_inner instead.
+
+    """
+    result = 0
+    for sentence in sentences:
+        word_vocabs = [model.wv.vocab[w] for w in sentence if w in model.wv.vocab and
+                       model.wv.vocab[w].sample_int > model.random.rand() * 2**32]
+        for pos, word in enumerate(word_vocabs):
+            reduced_window = model.random.randint(model.window)  # `b` in the original word2vec code
+
+            # subword_indices = []
+            # subword_indices += get_subwords(word)
+
+            # now go over all words from the (reduced) window, predicting each one in turn
+            start = max(0, pos - model.window + reduced_window)
+            for pos2, word2 in enumerate(word_vocabs[start:(pos + model.window + 1 - reduced_window)], start):
+                # don't train on the `word` itself
+
+                subword2_indices = []
+            	subword2_indices += get_subwords(word2)
+
+                if pos2 != pos:
+                	train_sg_pair(model, model.wv.index2word[word.index], subword2_indices, alpha)
+                    # train_sg_pair(model, model.wv.index2word[word.index], word2.index, alpha)
+
+                    # isn't model.wv.index2word[word.index] same as word ??
+        result += len(word_vocabs)
+    return result
+
+def train_batch_cbow(model, sentences, alpha, work=None, neu1=None):
+	"""
+	Update CBOW model by training on a sequence of sentences.
+
+	"""
+	result = 0
+	for sentence in sentences:
+		# word_vocabs bow in fasttext.cc
+		word_vocabs = [model.wv.vocab[w] for w in sentence if w in model.wv.vocab and
+					   model.wv.vocab[w].sample_int > model.random.rand() * 2**32]
+		for pos, word in enumerate(word_vocabs):
+			reduced_window = model.random.randint(model.window)  # `b` in the original word2vec code  ## why random
+			start = max(0, pos - model.window + reduced_window)
+			window_pos = enumerate(word_vocabs[start:(pos + model.window + 1 - reduced_window)], start)
+			word2_indices = [word2.index for pos2, word2 in window_pos if (word2 is not None and pos2 != pos)]
+			# l1 = np_sum(model.wv.syn0[word2_indices], axis=0)  # 1 x vector_size
+			# if word2_indices and model.cbow_mean:
+			#	l1 /= len(word2_indices)
+
+			# for loop for ngrams for words in this window here
+			word2_subwords_indices = []
+
+			for indices in word2_indices:
+				word2_subwords_indices += get_subwords(model.wv.syn0[indices])  # subwords for each word in window except target word
+
+				# probably, we need to put subwords in model.wv.syn0 too
+
+			l1 = np_sum(model.wv.syn0[word2_subwords_indices], axis=0)  # 1 x vector_size
```

Comment:
I don't think storing ngram weights in `syn0` is a good idea, because it makes it much harder to use methods from `KeyedVectors`. We should have a separate matrix for ngram weights (maybe `syn0_all`, like in the `FastTextKeyedVectors` from the wrapper, to be able to reuse that class for methods like `most_similar` etc).
Once training is complete, `syn0` should be populated with computed in-vocab-word vectors, to not have to recompute word vectors for in-vocab words on every lookup.